### PR TITLE
compute shaders v2

### PR DESCRIPTION
**Pull request by haasn**
Refactored and modernized my code from X years ago. Mostly this is just to add compute shader support, so I can then start rewriting e.g. EWA samplers or user shaders using them.

But to prevent the code from sitting around unused for now, I decided to try my hand at HDR signal peak detection. Comments welcome

### CODE REVIEW

**Comment by haasn** [comment]
Changed my mind about the compute shader API, instead I just merged the right dispatch into `finish_pass_fbo`. So for the code there's no longer a difference between `finish_pass_fbo` and `finish_pass_compute`, it just depends on whether or not you set `p->compute_w/h`. Now the code integrates exceptionally well, there's barely a “user” (i.e. video.c)-facing difference between fragment and compute shaders. This will also make it trivial to support compute shaders as e.g. custom hooks without any more metadata. Overall I'm much happier with this commit (29134e2) now.

I've also changed my mind about HDR peak detection: Instead of hackily doing it in linear light during the scaling step, we should just do it during the output step. The reason for this is simple: All HDR videos are 4K anyway, so they will basically never be upscaled - if anything, downscaled. So doing it after scaling is actually usually a *good* idea, rather than a bad idea. (And for 4K monitors it doesn't make a difference)

Besides, doing it at 4K is so cheap anyway. So instead, I would like to move the compute shader for this to the place where it belongs (`pass_tone_map`), which unfortunately seems to require integrating SSBOs with the gl_shader_cache. So there's no way around that. Oh well, I'll figure it out.

**Comment by haasn** [comment]
> Besides, doing it at 4K is so cheap anyway. So instead, I would like to move the compute shader for this to the place where it belongs (pass_tone_map), which unfortunately seems to require integrating SSBOs with the gl_shader_cache. So there's no way around that. Oh well, I'll figure it out.

Actually, there might be something tricky about this approach: If we use a compute scaler (e.g. `pass_compute_polar`) at the same time as HDR peak detection, they would both technically be running inside the same pass (except when using interpolation). So in this case, the question arises: what block size do we use?

For example the compute scaler will most likely have some fairly precise block size requirements, but the HDR peak shader prefers to run in 8x8 if it can (since that seems to be faster than 16x16 and 32x32).

I guess I could just do something simple like always picking the greater of the two block sizes, but that feels... I dunno. Might cause us some nasty surprises later on; but it's not like we can avoid the issue of running multiple computer shaders in general.

(And actually the compute code won't have _that_ strict size requirements, it just needs to be at least 32 horizontally - that's the only concern)

**Comment by haasn** [comment]
Hmm, there are actually more issues with this: If we directly output to the FBO (`finish_pass_direct`), we can't use compute shaders, because those can't draw to FBOs. So we need to use an extra indirection for this case.

**Comment by haasn** [comment]
Updated; fixed the texcoord emulation, improved the API some more, moved the peak code to the right place.

Now it's simpler and requires fewer layering violations. The big one right now that I would love to somehow get rid of is the SSBO.

**Comment by haasn** [comment]
Added support for EWA samplers (polar). This is a massive speedup:

git master:
git master
computev2:
[!computev2](https://snow.haasn.xyz/~nand/mpv-shot0002.jpg)

**Comment by haasn** [comment]
@bjin @igv User compute shaders are now supported, in case you want to play around with it. I'll write a few examples whenever I feel like it.

**Comment by haasn** [comment]
Hmm, there's actually an annoying issue with the user compute shaders: since the `texcoord` etc. get added in `finish_pass_fbo`, but the `vec4 hook() { }` from the user shader got appended to the header before that, we can't actually reference `texcoord0` in the user shader (yet). No idea what a clean work-around would be. There's no easy way to “prepend” to a bstr, and it feels hacky to add like a `header2`. But it would also be difficult to get the user shader loaded *after* the `finish_pass_fbo`. Another approach would be to pass the compute shader metadata to `gl_sc_generate` so it can introduce the vertices on its own, but that's also annoying.

It should still be possible to test them using `texelFetch` directly, though.

**Comment by haasn** [comment]
Here's an example. Due to the above problem, I had to hard-code `wsize` instead of using the (normally dynamic) `w_sz`. But otherwise it works fine:

```glsl
//!HOOK LUMA
//!BIND HOOKED
//!COMPUTE 32 16
//!DESC 9x9 average

// fixme: this would be w_sz normally
const ivec2 wsize = ivec2(32, 16);

// needs extra 4x4 of padding on all sides, because the 9x9 kernel will have
// some overhang outside of the working group
const ivec2 padding = ivec2(4, 4);
const ivec2 isize = ivec2(wsize) + 2*padding;

shared float inp[isize.y][isize.x];

vec4 hook()
{
    // load texels into shmem
    ivec2 base = ivec2(w_id) * wsize;
    for (uint y = l_id.y; y < isize.y; y += wsize.y)
        for (uint x = l_id.x; x < isize.x; x+= wsize.x)
            inp[y][x] = texelFetch(HOOKED_raw, base + ivec2(x,y) - padding, 0).x;

    // ensure writes visible
    groupMemoryBarrier();
    barrier();

    // do 9x9 convolution
    float sum;
    for (uint y = 0; y < 9; y++)
        for (uint x = 0; x < 9; x++)
            sum += inp[l_id.y+y][l_id.x+x];

    return vec4(sum / (9*9), 0, 0, 1);
}
```

This should implement how to do a very generic 1:1 interpolation kernel, the kind you'd see in a neural network like FSRCNN. Should be much, much faster than the equivalent texture() boilerplate.

**Comment by bjin** [comment]
A few questions about usage of compute shader.
1. What's the best way to load kernel weights from user texture into a global float array, which is later visible to all working groups?
2. How's the threads dispatched? My guess is `wsize.x* wsize.y` threads are dispatched for each block, and blocks are processed one after another, right? And this implies that barrier affects only local group?
3. Is it guaranteed that the actual block size is exactly the same as specified? If I want to process the whole image in single thread (for example, for error diffusion), what's the proper way to do it without dispatching unnecessary threads?
4. Is it possible to return a float/vec4 array as a whole as target texture?

**Comment by haasn** [comment]
> What's the best way to load kernel weights from user texture into a global float array, which is later visible to all working groups?

Global, to *all* working groups? That would have to either be a UBO, an SSBO or a texture - of which only the latter would currently be supported (with the usertex branch). But a word of caution: if you're going to be sampling from these weights dynamically, then you're destroying any benefit you might get from the UBO. Just use a texture. And if you're loading from them in a 1:1 relationship, you can unroll the loop and hard-code the weights.

> How's the threads dispatched? My guess is wsize.x* wsize.y threads are dispatched for each block, and blocks are processed one after another, right? And this implies that barrier affects only local group?

Standard compute shader semantics. Each block has `wsize.x * wsize.y` threads, and these threads are further subdivided into warps (of size 32 on nvidia, and I believe 64 on AMD). Working groups run independently, but in any (unspecified) order. `barrier()` only synchronizes threads within a working group, and `groupMemoryBarrier()` only affects writes within the working group's shared memory.

> Is it guaranteed that the actual block size is exactly the same as specified?

Yes

> If I want to process the whole image in single thread 

You don't, that would be impossibly slow. You need to have big enough working groups (e.g. 256 threads) to leverage parallelism within them.

**Comment by bjin** [comment]
It seems that compute shader could reduce all those redundant texel fetch from same working group considerably. But just like the `textureGatherOffset` trick, it probably won't help much with NNEDI3, since NNEDI3 is computation intensive. Rather, it will probably improve the performance of RAVU a lot, along with applied `textureGatherOffset` trick.

How's the `texcoord` issue by the way? It seems to be required to use `textureGatherOffset`.

**Comment by haasn** [comment]
Re: textureGatherOffset, for EWA compute shaders I found that it decreases performance, rather than increases it. Reason is simple:

You have 256 parallel threads (32x8), so that means 256 threads can fetch things at the same time. But for something like 2x upscaling, you only need 23x11 source texels for this same working group, which is 253 items.

So in other words, you can optimally load all 256 source texels into the array in a single texture() call per thread, which is extremely fast and efficient. If you're doing this with textureGatherOffset on the other hand, you have the folowing issues:

1. 23x11 is odd, so you end up doing redundant work by loading 24x12
2. 24*12 pixels / (4 pixels/fetch) = 52 texels, so the other 256-52 = 204 threads are sleeping doing absolutely nothing
3. Each thread that *does* something also has to do more (1. loop over components, and 2. loop over four pixels and do four shmem writes)
4. The threads that *are* active run into bank conflicts, because the *2 scaling means you're only accessing every other array element at the same time, thus decreasing performance

tl;dr it's really not worth it to use textureGatherOffset if you're going to be sharing all of the results anyway, just do one texture() per thread which is extremely fast. Even better might be one texelFetch per thread (which I'll benchmark once I figure out how to resolve the `texcoord` issue)

**Comment by haasn** [comment]
Thought up a solution for the texcoord problem, I'll try implementing it when I next have free time. (I haven't slept properly since tuesday, and I'm supposed to do work on the weekend for my thesis, so it might be a few days)

**Comment by haasn** [comment]
My idea for texcoords didn't work out, so I just went with the ugly approach that I was trying to avoid (introducing a second level of “headers”). Apparently somebody else thought of this solution in the past, though, since the field `struct bstr prelude_text` already existed in the sc, even though it was unused... I just repurposed that, little harm done.

Anyway, I've adapted the convolution example to now be fully dynamic:

```glsl
//!HOOK LUMA
//!BIND HOOKED
//!COMPUTE 32 8
//!DESC avg convolution

// Kernel size, 9x9 as an example
const ivec2 ksize = ivec2(9, 9);
const ivec2 offset = ksize / 2;

// We need to load extra source texels to account for padding due to kernel
// overhang
const ivec2 isize = ivec2(w_sz) + ksize - 1;

shared float inp[isize.y][isize.x];

vec4 hook()
{
    // load texels into shmem
    ivec2 base = ivec2(w_id) * ivec2(w_sz);
    for (uint y = l_id.y; y < isize.y; y += w_sz.y) {
        for (uint x = l_id.x; x < isize.x; x += w_sz.x)
            inp[y][x] = texelFetch(HOOKED_raw, base + ivec2(x,y) - offset, 0).x;
    }

    // ensure writes visible
    groupMemoryBarrier();
    barrier();

    // do convolution
    float sum;
    for (uint y = 0; y < ksize.y; y++) {
        for (uint x = 0; x < ksize.x; x++)
            sum += inp[l_id.y+y][l_id.x+x];
    }

    return vec4(sum / (ksize.x * ksize.y), 0, 0, 1);
}
```

**Comment by haasn** [comment]
So this is pretty much ready for merging. Anybody have objections before we do so?

**Comment by haasn** [comment]
Merged

**Comment by bjin** [comment]
@haasn I'm currently looking into the possibility to port RAVU to compute shader. I have a few questions.
1. You mentioned earlier that `shared vec4[]` is very slow, do you know any workaround? Just like `superxbr-native`, I would like to have RAVU able to upscale combined planes (convolution kernel extracted based on luma, but applies to `vec4` later).
2. You also mentioned that compute shader would be able to modify SSBO or even texture directly. How it works? Is it same as current invocation conversion just without return value. And a specified SSBO array or a modifiable  texture will later be saved as texture?

**Comment by haasn** [comment]
@bjin The EWA shader has the same problem, and the solution is simple. Instead of one 'shared vec4' it uses this:

```glsl
[vo/opengl] [  1] shared float in0[11][23];
[vo/opengl] [  2] shared float in1[11][23];
[vo/opengl] [  3] shared float in2[11][23];
```

And then when loading the samples:
```glsl
[vo/opengl] [ 15] for (int y = int(gl_LocalInvocationID.y); y < 11; y += 8) {
[vo/opengl] [ 16] for (int x = int(gl_LocalInvocationID.x); x < 23; x += 32) {
[vo/opengl] [ 17] c = texture(tex, wbase + pt * vec2(x - 2, y - 2));
[vo/opengl] [ 18] in0[y][x] = c[0];
[vo/opengl] [ 19] in1[y][x] = c[1];
[vo/opengl] [ 20] in2[y][x] = c[2];
[vo/opengl] [ 21] }}
[vo/opengl] [ 22] groupMemoryBarrier();
[vo/opengl] [ 23] barrier();
```

Basically this construction reorders it from “R G B X R G B X R G R G B X ...” to “R R R R ... G G G G ... B B B B ....”, which is better because now all of the ‘R’ pixels will be accessed simultaneously, then all of the ‘G’ pixels, and then all of the ‘B’ pixels - and all of these simultaneous accessess are offset by exactly one `float` (address in memory).

Doing it the vec4 way is slow because accessing all of the ‘R’ vectors simultaneously basically means you're accessing offset 0, offset 4, offset 8, offset 12 and so on at the same time - which collides due to bank conflicts. It's much better to access offset 0, offset 1, offset 2, offset 3 and so on at the same time.

> You also mentioned that compute shader would be able to modify SSBO or even texture directly. How it works? Is it same as current invocation conversion just without return value.

Not currently possible for user compute shaders. It's just a theoretical idea. The way it works is:

1. Bind an image or SSBO for writing
2. Modify it in the compute shader
3. Insert a glMemoryBarrier() call
4. Attach it as a texture or SSBO for reading in another shader
